<!DOCTYPE html>
<html  lang="zh">
<head>
    <meta name="google-site-verification" content="GfJLBEu3BHvdazYMXsGETTsY_cx2FLLXrqcx0H0s4cA" />
    <meta charset="utf-8" />

<meta name="generator" content="Hexo 4.2.0" />

<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1" />

<title>callback, promise, async/await 使用方式教學以及介紹 Part II (Error Handling 介紹) - 技術雜記 Technology Notes - Jack Yu | 傑克</title>

<script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
})(window,document,'script','dataLayer','GTM-NKT99TP');</script>

<script data-ad-client="ca-pub-1515253208026743" async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>


    <meta name="description" content="上一篇主要是介紹如何使用這篇會介紹該如何去在每一種使用方式之中去做 Error Handling">
<meta property="og:type" content="article">
<meta property="og:title" content="callback, promise, async&#x2F;await 使用方式教學以及介紹 Part II (Error Handling 介紹)">
<meta property="og:url" content="https://yu-jack.github.io/2019/05/02/promise-2/index.html">
<meta property="og:site_name" content="技術雜記 Technology Notes - Jack Yu | 傑克">
<meta property="og:description" content="上一篇主要是介紹如何使用這篇會介紹該如何去在每一種使用方式之中去做 Error Handling">
<meta property="og:locale" content="zh_TW">
<meta property="og:image" content="https://yu-jack.github.io/images/og_image.png">
<meta property="article:published_time" content="2019-05-02T14:42:27.000Z">
<meta property="article:modified_time" content="2020-01-29T08:25:29.000Z">
<meta property="article:author" content="Jack Yu">
<meta property="article:tag" content="nodejs">
<meta property="article:tag" content="JavaScript">
<meta property="article:tag" content="promise">
<meta property="article:tag" content="async">
<meta property="article:tag" content="await">
<meta property="article:tag" content="callback">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://yu-jack.github.io/images/og_image.png">







<link rel="icon" href="/images/favicon.svg">


<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bulma@0.7.2/css/bulma.css">
<link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.4.1/css/all.css">
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Ubuntu:400,600|Source+Code+Pro">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/highlight.js@9.12.0/styles/atom-one-dark.css">

    
    
<style>body>.footer,body>.navbar,body>.section{opacity:0}</style>

    
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/lightgallery@1.6.8/dist/css/lightgallery.min.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/justifiedGallery@3.7.0/dist/css/justifiedGallery.min.css">

    
    
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/outdatedbrowser@1.1.5/outdatedbrowser/outdatedbrowser.min.css">

    
    
    
    
<link rel="stylesheet" href="/css/back-to-top.css">

    
    
    
    
    
    
    
    <link rel="stylesheet" href="/css/progressbar.css">
<script src="https://cdn.jsdelivr.net/npm/pace-js@1.0.2/pace.min.js"></script>
    
    
    


<link rel="stylesheet" href="/css/style.css">
    <style>
        .content img {
            display: block;   
            margin: 0 auto;
        }
    </style>
</head>
<body class="is-2-column">
    <nav class="navbar navbar-main">
    <div class="container">
        <!-- <div class="navbar-brand is-flex-center">
            <a class="navbar-item navbar-logo" href="/">
            
                <img src="/images/logo.svg" alt="callback, promise, async/await 使用方式教學以及介紹 Part II (Error Handling 介紹)" height="28">
            
            </a>
        </div> -->
        <div class="navbar-menu">
            
            <div class="navbar-start">
                
                <a class="navbar-item"
                href="/">Home</a>
                
                <a class="navbar-item"
                href="/archives">Archives</a>
                
                <a class="navbar-item"
                href="/categories">Categories</a>
                
                <a class="navbar-item"
                href="/tags">Tags</a>
                
            </div>
            
            <div class="navbar-end">
                
                
                <a class="navbar-item is-hidden-tablet catalogue" title="文章目錄" href="javascript:;">
                    <i class="fas fa-list-ul"></i>
                </a>
                
                
                <a class="navbar-item search" title="搜尋" href="javascript:;">
                    <i class="fas fa-search"></i>
                </a>
                
            </div>
        </div>
    </div>
</nav>
    
    <section class="section">
        <div class="container">
            <div class="columns">
                <div class="column is-8-tablet is-8-desktop is-8-widescreen has-order-2 column-main">
<div class="card">
    
    <div class="card-content article ">
        
        <div class="level article-meta is-size-7 is-uppercase is-mobile is-overflow-x-auto">
            <div class="level-left">
                <time class="level-item has-text-grey" datetime="2019-05-02T14:42:27.000Z">2019-05-02</time>
                
                <div class="level-item">
                <a class="has-link-grey -link" href="/categories/JavaScript/">JavaScript</a>
                </div>
                
                
                <span class="level-item has-text-grey">
                    
                    
                    11 分鐘 閱讀文 (大約 1625 個字)
                </span>
                
                
                
                <span id="busuanzi_container_page_pv">
                    <span id="busuanzi_value_page_pv"></span> 觀看次數
                </span>
                
            </div>
        </div>
        
        <h1 class="title is-size-3 is-size-4-mobile has-text-weight-normal">
            
                callback, promise, async/await 使用方式教學以及介紹 Part II (Error Handling 介紹)
            
        </h1>
        
            <div class="content" style="word-break: break-all;">
                <p>上一篇主要是介紹如何使用<br>這篇會介紹該如何去在每一種使用方式之中去做 Error Handling  </p>
<a id="more"></a>
<h2><span id="callback">callback</span></h2><p>相信各位有在使用別人第三方套件或是 Node.js 原生的 Library 都會發現一件事情<br>那就是 callback 第一個參數都會是 error<br>雖然這看似是一個不成文的規定，但仔細想想把 error 放在第一個是非常合理的<br>假設當 callback 參數回傳越來越多的時候，總不可能把 error 放在最後一個去處理<br>因為你會始終不知道哪一個會是 error (就算寫註解也會讀到瘋掉)<br>試想一下這幾段 code 就可以理解了</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">test</span>(<span class="params">cb</span>) </span>&#123;</span><br><span class="line">    <span class="comment">// 當 function 成功後</span></span><br><span class="line">    cb(successful_data_1, successful_data_2)</span><br><span class="line">&#125;</span><br><span class="line">test(<span class="function">(<span class="params">successful_data_1, successful_data_2</span>) =&gt;</span> &#123;</span><br><span class="line">    <span class="comment">// 開心地處理兩個回傳的資料</span></span><br><span class="line">&#125;)</span><br><span class="line"><span class="comment">// --- 分隔線  </span></span><br><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">test</span>(<span class="params">cb</span>) </span>&#123;</span><br><span class="line">    <span class="comment">// 當 function 失敗後</span></span><br><span class="line">    cb(error)</span><br><span class="line">&#125;</span><br><span class="line">test(<span class="function">(<span class="params">error</span>) =&gt;</span> &#123;</span><br><span class="line">    <span class="comment">// 咦? 第一個到底是 error 還是我原本的 successful_data_1</span></span><br><span class="line">&#125;)</span><br></pre></td></tr></table></figure>
<p>當遇到上面的狀況就會變得非常難判斷，但如果我整體改寫成這樣就會變得輕而易舉  </p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">test</span>(<span class="params">cb</span>) </span>&#123;</span><br><span class="line">    <span class="comment">// 當 function 成功後</span></span><br><span class="line">    cb(<span class="literal">null</span>, successful_data_1, successful_data_2)</span><br><span class="line">&#125;</span><br><span class="line">test(<span class="function">(<span class="params">error, successful_data_1, successful_data_2</span>) =&gt;</span> &#123;</span><br><span class="line">    <span class="keyword">if</span> (error != <span class="literal">null</span>) &#123;</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="comment">// 開心地處理兩個回傳的資料</span></span><br><span class="line">&#125;)</span><br><span class="line"><span class="comment">// --- 分隔線  </span></span><br><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">test</span>(<span class="params">cb</span>) </span>&#123;</span><br><span class="line">    <span class="comment">// 當 function 失敗後</span></span><br><span class="line">    cb(error)</span><br><span class="line">&#125;</span><br><span class="line">test(<span class="function">(<span class="params">error, successful_data_1, successful_data_2</span>) =&gt;</span> &#123;</span><br><span class="line">    <span class="keyword">if</span> (error != <span class="literal">null</span>) &#123;</span><br><span class="line">        <span class="comment">// 開心地處理 error, 於是 data_1 以及 data_2 就完全不用管他們了</span></span><br><span class="line">    &#125;</span><br><span class="line">&#125;)</span><br></pre></td></tr></table></figure>
<p>當然有人會說<br>『啊我就把所有參數丟到第一個當 Object 全部存起來，第二個就放 Error 也是一種方式啊』<br>這樣講的話當然沒錯，但如果把所有東西都放在第一個 Object 裡面<br>這樣參數就會有分類，使用問題也只是會徒增而已<br>再加上這算是一種共識了，所以跟潛規則走會比較方便一點  </p>
<h2><span id="promise">promise</span></h2><p>Promise 處理 error 的方式就比較特別了，我們先來看看一般 promise 出錯的時候是怎麼抓取的  </p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">test</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">return</span> <span class="keyword">new</span> <span class="built_in">Promise</span>(<span class="function">(<span class="params">res, rej</span>) =&gt;</span> &#123;</span><br><span class="line">        rej(<span class="string">"this is error"</span>);</span><br><span class="line">    &#125;)</span><br><span class="line">&#125;</span><br><span class="line">test().then(<span class="function">(<span class="params">data</span>) =&gt;</span> &#123;<span class="built_in">console</span>.log(<span class="string">"Get "</span> + data)&#125;).catch(<span class="function">(<span class="params">error</span>) =&gt;</span> &#123;<span class="built_in">console</span>.log(<span class="string">"handle! "</span> + error)&#125;); <span class="comment">// this is error</span></span><br></pre></td></tr></table></figure>
<p>上面為一般 promise 用 rej 的方式，外面用 catch 去抓住這個錯誤<br>但凡事要考慮例外，萬一有一個 error 是你沒辦法 rej 到的話，那該要怎麼抓取?  </p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">test</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">return</span> <span class="keyword">new</span> <span class="built_in">Promise</span>(<span class="function">(<span class="params">res, rej</span>) =&gt;</span> &#123;</span><br><span class="line">        oqiwje() <span class="comment">// non-exist function</span></span><br><span class="line">    &#125;)</span><br><span class="line">&#125;</span><br><span class="line">test().then(<span class="function">(<span class="params">data</span>) =&gt;</span> &#123;<span class="built_in">console</span>.log(<span class="string">"Get "</span> + data)&#125;).catch(<span class="function">(<span class="params">error</span>) =&gt;</span> &#123;<span class="built_in">console</span>.log(<span class="string">"handle! "</span> + error)&#125;); <span class="comment">// this is error</span></span><br></pre></td></tr></table></figure>
<p>會發現當在 Promise 裡面出錯的話，外面的 catch 也是能抓到的<br>其原因是因為 Promise 是有被一層內部的 try-catch 給包住<br>且在內部的 catch 那一邊套用了預設地 rej function<br>所以外面才抓得到</p>
<p>那如果放在 Promise 外面的話呢??  </p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">test</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    oqiwje() <span class="comment">// non-exist function</span></span><br><span class="line">    <span class="keyword">return</span> <span class="keyword">new</span> <span class="built_in">Promise</span>(<span class="function">(<span class="params">res, rej</span>) =&gt;</span> &#123;</span><br><span class="line">    &#125;)</span><br><span class="line">&#125;</span><br><span class="line">test().then(<span class="function">(<span class="params">data</span>) =&gt;</span> &#123;<span class="built_in">console</span>.log(<span class="string">"Get "</span> + data)&#125;).catch(<span class="function">(<span class="params">error</span>) =&gt;</span> &#123;<span class="built_in">console</span>.log(<span class="string">"handle! "</span> + error)&#125;);</span><br></pre></td></tr></table></figure>
<p>咦!? 竟然抓不到，error 直接噴出來!<br>但這也不意外，因為出了 Promise 到了外面<br>那就是要透過自己去寫 try-catch 才能抓取到這個錯誤  </p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">test</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    oqiwje() <span class="comment">// non-exist function</span></span><br><span class="line">    <span class="keyword">return</span> <span class="keyword">new</span> <span class="built_in">Promise</span>(<span class="function">(<span class="params">res, rej</span>) =&gt;</span> &#123;</span><br><span class="line">    &#125;)</span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line">    test().then(<span class="function">(<span class="params">data</span>) =&gt;</span> &#123;<span class="built_in">console</span>.log(<span class="string">"Get "</span> + data)&#125;).catch(<span class="function">(<span class="params">error</span>) =&gt;</span> &#123;<span class="built_in">console</span>.log(<span class="string">"handle! "</span> + error)&#125;);</span><br><span class="line">&#125; <span class="keyword">catch</span> (error) &#123;</span><br><span class="line">    <span class="built_in">console</span>.log(<span class="string">"handled by outer try-catch"</span>);</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>還有一種 Handle 方式是寫在內層 function 裡面</p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">test</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">return</span> <span class="keyword">new</span> <span class="built_in">Promise</span>(<span class="function">(<span class="params">res, rej</span>) =&gt;</span> &#123;</span><br><span class="line">        oqiwje() <span class="comment">// non-exist function</span></span><br><span class="line">    &#125;).catch(<span class="function"><span class="params">error</span> =&gt;</span> <span class="string">"handle by inner function"</span>)</span><br><span class="line">&#125;</span><br><span class="line">test().then(<span class="function">(<span class="params">data</span>) =&gt;</span> &#123;<span class="built_in">console</span>.log(<span class="string">"Get "</span> + data)&#125;).catch(<span class="function">(<span class="params">error</span>) =&gt;</span> &#123;<span class="built_in">console</span>.log(<span class="string">"handle! "</span> + error)&#125;); <span class="comment">// Get handle by inner function</span></span><br></pre></td></tr></table></figure>
<p>那因為在 inner function 裡面被抓取到，並且回傳<br>還記得 promise chain 中，如果 return 的話是會到下一個 then 去的<br>所以這邊會被外面的 then 給抓到，而不是 catch，這邊要注意  </p>
<p>Promise 的 Error Handling 只要能確保能執行到 rej 就沒什麼問題了<br>然而在 Promise 之前用 try-catch 包起來或程式都丟到 Promise 裡面等他報錯丟出來也可以處理到  </p>
<h2><span id="asyncawait">async/await</span></h2><p>await catch error 的方式可以想成一般 try-catch 的方式  </p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">test</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">return</span> <span class="keyword">new</span> <span class="built_in">Promise</span>(<span class="function">(<span class="params">res, rej</span>) =&gt;</span> &#123;</span><br><span class="line">        rej(<span class="string">"QQQ"</span>);</span><br><span class="line">    &#125;);</span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">async</span> <span class="function"><span class="keyword">function</span> <span class="title">main</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        <span class="keyword">let</span> result = <span class="keyword">await</span> test()</span><br><span class="line">    &#125; <span class="keyword">catch</span> (error) &#123;</span><br><span class="line">        <span class="built_in">console</span>.log(<span class="string">"Handled by main"</span>)</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">main()</span><br></pre></td></tr></table></figure>
<p>而要特別注意的是，如果把 catch 寫在外面的 await 那裡的話<br>會造成程式不會往最外層的 catch 前進  </p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">test</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">return</span> <span class="keyword">new</span> <span class="built_in">Promise</span>(<span class="function">(<span class="params">res, rej</span>) =&gt;</span> &#123;</span><br><span class="line">        rej(<span class="string">"QQQ"</span>);</span><br><span class="line">    &#125;);</span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">async</span> <span class="function"><span class="keyword">function</span> <span class="title">main</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        <span class="keyword">let</span> result = <span class="keyword">await</span> test().catch(<span class="function"><span class="params">()</span> =&gt;</span> &#123;<span class="built_in">console</span>.log(<span class="string">"Handled by await"</span>)&#125;)</span><br><span class="line">        <span class="comment">// 因為有正確被 handle 到，所以程式是會繼續下去執行的</span></span><br><span class="line">        <span class="built_in">console</span>.log(<span class="string">"Still going"</span>)</span><br><span class="line">    &#125; <span class="keyword">catch</span> (error) &#123;</span><br><span class="line">        <span class="built_in">console</span>.log(<span class="string">"Handled by main"</span>)</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">main()</span><br></pre></td></tr></table></figure>
<p>但如果透過在 catch function 裡面把 error 再次 throw 出來的話，是可以成功 throw 出來  </p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">test</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">return</span> <span class="keyword">new</span> <span class="built_in">Promise</span>(<span class="function">(<span class="params">res, rej</span>) =&gt;</span> &#123;</span><br><span class="line">        rej(<span class="string">"QQQ"</span>);</span><br><span class="line">    &#125;);</span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">async</span> <span class="function"><span class="keyword">function</span> <span class="title">main</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        <span class="keyword">let</span> result = <span class="keyword">await</span> test().catch(<span class="function"><span class="params">()</span> =&gt;</span> &#123;</span><br><span class="line">            <span class="built_in">console</span>.log(<span class="string">"Handled by await"</span>)</span><br><span class="line">            <span class="keyword">throw</span> <span class="keyword">new</span> <span class="built_in">Error</span>(<span class="string">"QQQQ"</span>)</span><br><span class="line">        &#125;)</span><br><span class="line">        <span class="comment">// 因為在上面做 throw error，所以程式不會繼續走下去</span></span><br><span class="line">        <span class="built_in">console</span>.log(<span class="string">"Still going"</span>)</span><br><span class="line">    &#125; <span class="keyword">catch</span> (error) &#123;</span><br><span class="line">        <span class="built_in">console</span>.log(<span class="string">"Handled by main"</span>)</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">main()</span><br></pre></td></tr></table></figure>
<p>千萬要注意 return 和 throw 的方式會帶來不一樣的結果<br>使用 return 就跟 Promise 的 reject 的狀態下 return 是一樣的<br>他會回傳到下一個 then 裡面 (也就是 resolved 的狀態)  </p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">test</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">return</span> <span class="keyword">new</span> <span class="built_in">Promise</span>(<span class="function">(<span class="params">res, rej</span>) =&gt;</span> &#123;</span><br><span class="line">        rej(<span class="string">"QQQ"</span>);</span><br><span class="line">    &#125;);</span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">async</span> <span class="function"><span class="keyword">function</span> <span class="title">main</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        <span class="keyword">let</span> result = <span class="keyword">await</span> test().catch(<span class="function"><span class="params">()</span> =&gt;</span> &#123;</span><br><span class="line">            <span class="built_in">console</span>.log(<span class="string">"Handled by await"</span>)</span><br><span class="line">            <span class="keyword">return</span> <span class="keyword">new</span> <span class="built_in">Error</span>(<span class="string">"QQQQ"</span>)</span><br><span class="line">        &#125;)</span><br><span class="line">        <span class="comment">// 因為在上面做 return, 相當於是把結果回傳到 result 裡面了</span></span><br><span class="line">        <span class="built_in">console</span>.log(result); <span class="comment">// Error: QQQQ</span></span><br><span class="line">        <span class="built_in">console</span>.log(<span class="string">"Still going"</span>)</span><br><span class="line">    &#125; <span class="keyword">catch</span> (error) &#123;</span><br><span class="line">        <span class="built_in">console</span>.log(<span class="string">"Handled by main"</span>)</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">main()</span><br></pre></td></tr></table></figure>
<p>而個人比較不建議的寫法是在 await 那一層做 Error Handling<br>而是盡量再底層那裡做 throw error 到最外面的 try-catch 去接<br>原因是這跟 Design Pattern 有關係<br>最外層的 main 可以想像是 Controller，而 test 可以想像成 Facade<br>在裡面得程式才是真正的商業邏輯<br>從下面程式來解讀的話，回家主要目的是要做功課<br>那做功課一定會有流程，像是先吃飯，洗澡，最後在讀書這樣的順序<br>但要怎麼吃飯洗澡讀書是要寫在每一個該做的項目的最裡面，而不會寫在順序那一層<br>這樣程式撰寫上會比較乾淨一點  </p>
<figure class="highlight javascript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">eatFirst</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">return</span> <span class="keyword">new</span> <span class="built_in">Promise</span>(<span class="function">(<span class="params">res, rej</span>) =&gt;</span> &#123;</span><br><span class="line">        setTimeout(<span class="function"><span class="params">()</span> =&gt;</span> &#123;</span><br><span class="line">            res(<span class="string">"Error"</span>);</span><br><span class="line">        &#125;, <span class="number">1000</span>);</span><br><span class="line">    &#125;);</span><br><span class="line">&#125;</span><br><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">getBook</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">return</span> <span class="keyword">new</span> <span class="built_in">Promise</span>(<span class="function">(<span class="params">res, rej</span>) =&gt;</span> &#123;</span><br><span class="line">        setTimeout(<span class="function"><span class="params">()</span> =&gt;</span> &#123;</span><br><span class="line">            res(<span class="string">"Error"</span>);</span><br><span class="line">        &#125;, <span class="number">1000</span>);</span><br><span class="line">    &#125;);</span><br><span class="line">&#125;</span><br><span class="line"><span class="function"><span class="keyword">function</span> <span class="title">writeIt</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">return</span> <span class="keyword">new</span> <span class="built_in">Promise</span>(<span class="function">(<span class="params">res, rej</span>) =&gt;</span> &#123;</span><br><span class="line">        setTimeout(<span class="function"><span class="params">()</span> =&gt;</span> &#123;</span><br><span class="line">            rej(<span class="string">"Books are ate by dogs!!!"</span>);</span><br><span class="line">        &#125;, <span class="number">1000</span>);</span><br><span class="line">    &#125;);</span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">async</span> <span class="function"><span class="keyword">function</span> <span class="title">doHomeWork</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">await</span> eatFirst()</span><br><span class="line">    <span class="keyword">await</span> getBook()</span><br><span class="line">    <span class="keyword">await</span> writeIt()</span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">async</span> <span class="function"><span class="keyword">function</span> <span class="title">main</span>(<span class="params"></span>) </span>&#123;</span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        <span class="keyword">let</span> result = <span class="keyword">await</span> doHomeWork();</span><br><span class="line">    &#125; <span class="keyword">catch</span> (error) &#123;</span><br><span class="line">        <span class="built_in">console</span>.log(<span class="string">"Handled by main"</span>)</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">main();</span><br></pre></td></tr></table></figure>
<h2><span id="後記">後記</span></h2><p>這次主要介紹 Error Handling 的方式<br>也加了一些個人建議撰寫的方法，如果有其他想法歡迎大家來討論！</p>

            </div>
        
        
        <div class="level is-size-7 is-uppercase">
            <div class="level-start">
                <div class="level-item">
                    <span class="is-size-6 has-text-grey has-mr-7">#</span>
                    <a class="has-link-grey -link" href="/tags/JavaScript/" rel="tag">JavaScript</a>, <a class="has-link-grey -link" href="/tags/async/" rel="tag">async</a>, <a class="has-link-grey -link" href="/tags/await/" rel="tag">await</a>, <a class="has-link-grey -link" href="/tags/callback/" rel="tag">callback</a>, <a class="has-link-grey -link" href="/tags/nodejs/" rel="tag">nodejs</a>, <a class="has-link-grey -link" href="/tags/promise/" rel="tag">promise</a>
                </div>
            </div>
        </div>
        
        
        
    </div>
</div>





<div class="card card-transparent">
    <div class="level post-navigation is-flex-wrap is-mobile">
        
        <div class="level-start">
            <a class="level level-item has-link-grey  article-nav-prev" href="/2019/06/02/ajax-with-session/">
                <i class="level-item fas fa-chevron-left"></i>
                <span class="level-item">前後端分離下之使用 session</span>
            </a>
        </div>
        
        
        <div class="level-end">
            <a class="level level-item has-link-grey  article-nav-next" href="/2019/04/24/javascript-this/">
                <span class="level-item">JavaScript this 是什麼? 如何運作的呢?</span>
                <i class="level-item fas fa-chevron-right"></i>
            </a>
        </div>
        
    </div>
</div>



<div class="card">
    <div class="card-content" style="display: flex; justify-content: center;" id="liker">
    </div>
    <script>
        var iframe = document.createElement("iframe")
        iframe.src = "https://button.like.co/in/embed/jk82421/button?referrer=" + encodeURIComponent(location.href)
        document.getElementById("liker").appendChild(iframe)
    </script>
    <div class="card-content">
        <!-- <h3 class="title is-5 has-text-weight-normal">評論</h3> -->
        
<script>
    var disqus_config = function () {
        this.page.url = 'https://yu-jack.github.io/2019/05/02/promise-2/';
        this.page.identifier = '2019/05/02/promise-2/';
    };
    (function() {
        var d = document, s = d.createElement('script');  
        s.src = '//' + 'yu-jack' + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>

<div id="disqus_thread">
    
    <noscript>Please enable JavaScript to view the <a href="//disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
</div>
    </div>
    <style>
        .custom-flex-container {
            display: flex; 
            justify-content: center; 
            margin-top: 30px;
        }
    </style>
</div>
</div>
                
    
    
    
    
    <!-- TODO: widgets 算錯 -->
    
    <div class="column is-4-tablet is-4-desktop is-3-widescreen  has-order-1 column-left is-sticky">
        
            
        
            
                

    <div class="card widget" id="toc">
        <div class="card-content" style="max-height: 85vh; overflow-y: auto;">
            <div class="menu">
                <h3 class="menu-label">
                    文章目錄
                </h3>
                <ul class="menu-list"><li>
        <a class="is-flex" href="#callback">
        <span class="has-mr-6">1</span>
        <span>callback</span>
        </a></li><li>
        <a class="is-flex" href="#promise">
        <span class="has-mr-6">2</span>
        <span>promise</span>
        </a></li><li>
        <a class="is-flex" href="#asyncawait">
        <span class="has-mr-6">3</span>
        <span>async/await</span>
        </a></li><li>
        <a class="is-flex" href="#後記">
        <span class="has-mr-6">4</span>
        <span>後記</span>
        </a></li></ul>
            </div>
        </div>
    </div>

            
        
            
        
            
        
        
            <div class="column-right-shadow is-hidden-widescreen ">
            
                
            
            </div>
        
    </div>
    

                
    
    
    
    
    <!-- TODO: widgets 算錯 -->
    
    <div class="column is-4-tablet is-4-desktop is-3-widescreen is-hidden-touch is-hidden-desktop-only has-order-3 column-right ">
        
            
        
        
    </div>
    

            </div>
        </div>
    </section>
    <footer class="footer">
    <div class="container">
        <div class="level">
            <div class="level-start has-text-centered-mobile">
                <p class="is-size-7">
                &copy; 2022 Jack Yu&nbsp;
                Powered by <a href="https://hexo.io/" target="_blank" rel="noopener">Hexo</a> & <a
                        href="https://github.com/ppoffice/hexo-theme-icarus" target="_blank" rel="noopener">Icarus</a> & Jack Yu
                
                </p>
            </div>
            <div class="level-end">
            
            </div>
        </div>
    </div>
</footer>
    <script src="https://cdn.jsdelivr.net/npm/jquery@3.3.1/dist/jquery.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/moment@2.22.2/min/moment-with-locales.min.js"></script>
<script>moment.locale("zh-TW");</script>


<script>
var IcarusThemeSettings = {
    site: {
        url: 'https://yu-jack.github.io',
        external_link: {"enable":true,"exclude":[]}
    },
    article: {
        highlight: {
            clipboard: true,
            fold: 'unfolded'
        }
    }
};
</script>


<script src="https://cdn.jsdelivr.net/npm/clipboard@2.0.4/dist/clipboard.min.js" defer></script>





<script src="/js/animation.js"></script>



<script src="https://cdn.jsdelivr.net/npm/lightgallery@1.6.8/dist/js/lightgallery.min.js" defer></script>
<script src="https://cdn.jsdelivr.net/npm/justifiedGallery@3.7.0/dist/js/jquery.justifiedGallery.min.js" defer></script>
<script src="/js/gallery.js" defer></script>



<div id="outdated">
    <h6>Your browser is out-of-date!</h6>
    <p>Update your browser to view this website correctly. <a id="btnUpdateBrowser" href="http://outdatedbrowser.com/">Update
            my browser now </a></p>
    <p class="last"><a href="#" id="btnCloseUpdateBrowser" title="Close">&times;</a></p>
</div>
<script src="https://cdn.jsdelivr.net/npm/outdatedbrowser@1.1.5/outdatedbrowser/outdatedbrowser.min.js" defer></script>
<script>
    document.addEventListener("DOMContentLoaded", function () {
        outdatedBrowser({
            bgColor: '#f25648',
            color: '#ffffff',
            lowerThan: 'flex'
        });
    });
</script>


<script src="https://cdn.jsdelivr.net/npm/mathjax@2.7.5/unpacked/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
<script>
document.addEventListener('DOMContentLoaded', function () {
    MathJax.Hub.Config({
        'HTML-CSS': {
            matchFontHeight: false
        },
        SVG: {
            matchFontHeight: false
        },
        CommonHTML: {
            matchFontHeight: false
        },
        tex2jax: {
            inlineMath: [
                ['$','$'],
                ['\\(','\\)']
            ]
        }
    });
});
</script>


<a id="back-to-top" title="回到頁首" href="javascript:;">
    <i class="fas fa-chevron-up"></i>
</a>
<script src="/js/back-to-top.js" defer></script>














<script src="/js/main.js" defer></script>

    
    <div class="searchbox ins-search">
    <div class="searchbox-container ins-search-container">
        <div class="searchbox-input-wrapper">
            <input type="text" class="searchbox-input ins-search-input" placeholder="請輸入關鍵字..." />
            <span class="searchbox-close ins-close ins-selectable"><i class="fa fa-times-circle"></i></span>
        </div>
        <div class="searchbox-result-wrapper ins-section-wrapper">
            <div class="ins-section-container"></div>
        </div>
    </div>
</div>
<script>
    (function (window) {
        var INSIGHT_CONFIG = {
            TRANSLATION: {
                POSTS: '文章',
                PAGES: '頁面',
                CATEGORIES: '分類',
                TAGS: '標籤',
                UNTITLED: '(無標題)',
            },
            CONTENT_URL: '/content.json',
        };
        window.INSIGHT_CONFIG = INSIGHT_CONFIG;
    })(window);
</script>
<script src="/js/insight.js" defer></script>
<link rel="stylesheet" href="/css/search.css">
<link rel="stylesheet" href="/css/insight.css">
    
    <script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</body>
</html>